### DISCUSSION TITLE
Custom Layout and Custom Interface Not Showing

### DISCUSSION THREAD

**Discussion by iandrianto** [directus/directus#16268] [Archive]
I tried to create a custom layout
https://docs.directus.io/extensions/layouts.html
my goal is creating a custom layout that can filter collection items based on a custom criteria

I put the files in `extensions/custom-layout/src/index.js` and `extensions/custom-layout/src/layout.vue`

Below is the  _package.json_ file

```json
{
  "name": "custom-layout",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "bundle": "npx rollup -c"
  },
  "devDependencies": {
    "@rollup/plugin-json": "^4.1.0",
    "@rollup/plugin-node-resolve": "^11.0.1",
    "@vue/compiler-sfc": "^3.0.5",
    "js-sha1": "^0.6.0",
    "rollup": "^2.36.1",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-import-css": "^1.1.4",
    "rollup-plugin-inject-process-env": "^1.3.1",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-terser": "^7.0.2",
    "rollup-plugin-vue": "^5.0.0",
    "vue-template-compiler": "^2.6.12"
  },
  "dependencies": {
    "mount-vue-component": "^0.9.5",
    "vue": "^2.6.12",
    "vue-file-agent": "^1.7.3"
  }
}
```

I've run `npm install` and `npm run bundle`
However after selecting it in the layout options, it only shows blank page.

When creating a custom interface following this article
https://docs.directus.io/extensions/interfaces.html
When selecting field type, the custom interface is not listed.

Not sure whether it's a bug or not
If not, please add a clear documentation and example how to build a custom layout and a custom interface (because the current documentation is very unclear and not understandable). or maybe add links to example projects that create custom Directus layout, interface, etc.

**Comment by azrikahar** [comment]
It's recommended to use the `create-directus-extension` CLI to create extensions as described here: https://docs.directus.io/extensions/creating-extensions.html as it'll include the `@directus/extensions-sdk` package which helps ease the config required to bundle your extension, as well as providing a starting template to work with.

Once you have built your extension, for example your custom-layout,  the _built_ file should be inserted into the extensions directory, not the source file. Here's a simplified example:

![](https://user-images.githubusercontent.com/42867097/199422367-c0ccb766-1fa1-4e01-bb24-e003c69c4a54.png)

In this case, the built `dist/index.js` file should be inserted to `<your directus project>/extensions/<extension type>/<your extension name/index.js` 👍

Additional reference: https://docs.directus.io/extensions/creating-extensions.html#local-extension

> If not, please add a clear documentation and example how to build a custom layout and a custom interface (because the current documentation is very unclear and not understandable)

We are actively looking to improve it, so do feel free to share any suggestions to enhance the docs in [`directus/docs`](https://github.com/directus/docs). 

> maybe add links to example projects that create custom Directus layout, interface, etc.

A great list of community extensions can be found here: https://github.com/directus-community/awesome-directus#extensions

**Comment by iandrianto** [comment]
Now it works after following your instruction

But another question about layout props. Is it possibile to add custom props other than listed on https://docs.directus.io/extensions/layouts.html

I tried this but the props value are not shown

_index.js_
```
import { ref } from 'vue';
import LayoutComponent from './layout.vue';
import { useApi, useStores } from '@directus/extensions-sdk';

export default {
	id: 'custom',
	name: 'Custom',
	icon: 'box',
	component: LayoutComponent,
	slots: {
		options: () => null,
		sidebar: () => null,
		actions: () => null,
	},
	async setup() {
		const name = ref('Custom Layout');
        const api = useApi();
		
		const response = await api.get('https://www.example.com');

		return { name, response, value:'value' };
	},
};
```

_layout.vue_
```
<template>
	<div>
		<p>Name: {{ name }}</p>
		<p>Collection: {{ collection }}</p>
		<p>Response: {{ response }}</p>
		<p>Value: {{ value }}</p>
	</div>
</template>

<script>
export default {
	inheritAttrs: false,
	props: {
		collection: {
			type: String,
			required: true,
		},
		name: {
			type: String,
			required: true,
		},
		response: {
			type: String,
			required: true,
		},
		value: {
			type: String,
			required: true,
		},
	},
};
</script>
```
